src/Language/DifferentialDatalog/OVSDB/Compile.hs

```diff
@@ -0,0 +1,127 @@
+    let (ColumnsProperty ovscols) : _ = tprops
+    uniqNames (\col -> "Multiple declarations of column " ++ col ++ " in table " ++ tableName) ovscols
+    columns <- mapM (mkCol isinput tableName) ovscols
+    let cols   = lefts  columns
+        tables = rights columns
+    let uuidcol = "_uuid: uuid"
+    let prefix = if isinput then "input" else empty
+    return $ prefix <+> "relation" <+> pp (name t) <> "("    $$
+             (nest' $ vcat $ punctuate comma (uuidcol:cols)) $$
+             ")"                                             $$
+             vcat tables
+
+mkCol :: (MonadError String me) => Bool -> String -> TableColumn -> me (Either Doc Doc)
+mkCol isinput tname c@TableColumn{..} = do
+    check (columnName /= "_uuid") (pos c) $ "Illegal column name _uuid in table " ++ tname
+    case columnType of
+         ColumnTypeAtomic at  -> (Left . (mkColName columnName <>) . (":" <+>)) <$> mkAtomicType at
+         ColumnTypeComplex ct -> mkComplexType isinput tname c ct
+
+mkColName :: String -> Doc
+mkColName "type" = "_type"
+mkColName "match" = "_match"
+mkColName n      = pp $ map toLower n
+
+mkAtomicType :: (MonadError String me) => AtomicType -> me Doc
+mkAtomicType IntegerType{} = return "integer"
+mkAtomicType (RealType p)  = err p "\"real\" types are not supported"
+mkAtomicType BooleanType{} = return "bool"
+mkAtomicType StringType{}  = return "string"
+mkAtomicType UUIDType{}    = return "uuid"
+
+mkComplexType :: (MonadError String me) => Bool -> String -> TableColumn -> ComplexType -> me (Either Doc Doc)
+mkComplexType isinput tname TableColumn{..} t@ComplexType{..} = do
+    let min = maybe 1 id minComplexType
+        max = maybe 1
+                    (\case
+                      Some x    -> x
+                      Unlimited -> fromIntegral (maxBound::Word64))
+                    maxComplexType
+    check (max >= min) (pos t) $ "min bound exceeds max bound"
+    check (min == 0 || min == 1) (pos t) $ "min bound must be 0 or 1"
+    check (max > 0) (pos t) $ "max bound must be greater than 0"
+    check (max /= 1 || isNothing valueComplexType) (pos t)
+          $ "Cannot handle key-value pairs when max bound is 1"
+    key <- mkBaseType keyComplexType
+    case (min, max) of
+         (1,1) -> return $ Left $ mkColName columnName <> ":" <+> key
+         (0,1) -> return $ Left $ mkColName columnName <> ": option_t<" <> key <> ">"
+         _     -> Right <$> mkSubtable isinput tname columnName t
+
+mkSubtable :: (MonadError String me) => Bool -> String -> String -> ComplexType -> me Doc
+mkSubtable isinput tname cname t@ComplexType{..} = do
+    keytype <- mkBaseType keyComplexType
+    valtype <- maybe (return []) (\v -> (\vt -> ["value:" <+> vt]) <$> mkBaseType v) valueComplexType
+    let cols = [ mkColName tname <> ": uuid"
+               , "key:" <+> keytype] ++ valtype
+    let prefix = if isinput then "input" else empty
+    return $ prefix <+> "relation" <+> pp tname <> "_" <> pp cname <> "(" $$
```

Comment:
Are you checking whether this name clashes with the existing ones?

Reply:
Nope, but DDlog will complain if it does.